components/script/script_thread.rs

```diff
@@ -1370,6 +1377,51 @@ impl ScriptThread {
         reports_chan.send(reports);
     }
 
+    /// Changes visibility of descendant iframes and slows down timers. Returns true if successful.
+    fn change_loaded_page_visibility(&self, id: PipelineId, visible: bool) -> bool {
+        if let Some(root_page) = self.page.borrow().as_ref() {
+            if let Some(ref inner_page) = root_page.find(id) {
+                let window = inner_page.window();
+                if visible {
+                    window.speed_up_timers();
+                } else {
+                    window.slow_down_timers();
+                }
+                let document = inner_page.document();
+                for iframe in &document.collect_descendant_iframes() {
+                    iframe.set_visible(visible);
```

Comment:
What about doing that at the pipeline level instead, where the frame tree is available.